### PR TITLE
BUG: fix an incompatibility with numpy 2.0 (np.lib.arraysetops is removed)

### PR DESCRIPTION
**Pull request by neutrinoceros**
### Description

fix e.g. `python -c "import astropy.coordinates"` with the latest and newest numpy dev.
for context: https://github.com/numpy/numpy/pull/24567

Opening as a draft because I'm not 100% confident my naive approach is the correct one yet.

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

**Comment by neutrinoceros** [comment]
tests with dev dependencies are passing so this is now ready for review

File: astropy/utils/masked/function_helpers.py
Comment (dhomeier) [inline astropy/utils/masked/function_helpers.py:140]:
```suggestion
    # public set operations from have been moved to the top-level namespace in numpy 2.0
    # (numpy/numpy#24507), raising an AttributeError when accessed through np.lib.arraysetops.
    # Listing them explicitly here.
```
Thanks for the fix; just some clarification to the comment (as to why they still need to be excluded at all).
Note that `numpy.lib.arraysetops` could still be imported explicitly, but from https://github.com/numpy/numpy/pull/24567#discussion_r1310266610 this looks like the right solution.

**Review by pllim — COMMENTED** [review]
Why not import from `np.lib._arraysetops_impl.__all__` like we do for the other stuff that got privatized?

**Comment by neutrinoceros** [comment]
That would work for me, though I do note that my current solution happens to be in line with recommendations from upstream. I'm happy to go with a different approach depending on consensus !

**Comment by pllim** [comment]
`units` and `utils.masked` use a lot of hidden stuff from numpy. This patch is not consistent with another similar patch from #15235 that @mhvk approved so we can wait for him to review if you want.

If we do it this way, we will have to also go back and change the other ones.

**Comment by dhomeier** [comment]
Something like https://github.com/astropy/astropy/blob/a0cd92c6ed8aa6a0bc8ff5e0bba596d5d50fbbe4/astropy/utils/masked/function_helpers.py#L1086-L1087 sure is a lot more concise!

File: astropy/utils/masked/function_helpers.py
Comment (mhvk) [inline astropy/utils/masked/function_helpers.py:139]: I'd prefer not to list them explicitly, but rather get them from the implementation. That way, if another function is added to their `__all__`, we'll notice (or in this case, know to continue to ignore it...).

I'm happy to revisit this after numpy 2.0 has actually appeared...

Comment (pllim) [reply astropy/utils/masked/function_helpers.py:139]: @neutrinoceros , are you able to make the requested change? Thank you! 🙏

File: astropy/utils/masked/function_helpers.py
Comment (dhomeier) [inline astropy/utils/masked/function_helpers.py:147]:
```suggestion
if NUMPY_LT_2_0:
    from np.lib import arraysetops
else:
    # Public set operations have been moved to the top-level namespace in numpy 2.0
    # (numpy/numpy#24507), raising an AttributeError when accessed through np.lib.arraysetops.
    from np.lib import _arraysetops_impl as arraysetops

IGNORED_FUNCTIONS |= {getattr(np, setopsname) for setopsname in arraysetops.__all__}
```

File: astropy/utils/masked/function_helpers.py
Comment (dhomeier) [reply astropy/utils/masked/function_helpers.py:139]: Actually we'd not notice as any new functions would automatically be added to `IGNORED_FUNCTIONS` while otherwise the CI would probably fail, pointing us to add it manually; but I think this way it's just as well.

File: astropy/utils/masked/function_helpers.py
Comment (neutrinoceros) [reply astropy/utils/masked/function_helpers.py:147]: @dhomeier I can't take your patch directly from the suggestion because `from np.lib ...` isn't valid (it should be `from numpy.lib ...`), but I'd be happy to push the correct version of it !
Other than that, it seems contradictory with [your other comment](https://github.com/astropy/astropy/pull/15253#discussion_r1310922130). Again, I'm happy to follow whatever consensus is reached.

**Comment by neutrinoceros** [comment]
I can confirm that CI fails when the union of implemented+ignored functions doesn't exactly match the whole set of functions that can be overloaded. I still think both approaches ( VS importing it from its new private location) are arguable, and both seem to have their own fragilities.
I propose we try a simple vote here using reaction to this comment:
- "🚀" for the current approach (hard coding the list of functions in that module as of numpy 1.25)
- "🎉" for the alternative approach (importing the list dynamically from its new private location) (https://github.com/astropy/astropy/pull/15253#discussion_r1310914575)

use both if you'd be happy either way (that's what I'm going to do) !

**Comment by dhomeier** [comment]
Thanks for investigating further, and sorry for the typo in suggestion 2. I agree with both as well ;-)
and think loading the private module would be legitimate as well, as we were already using a module that should have been private, just was not marked explicitly so, before.
Also the whole collection was set to get mask support at some point, per the original comment from c36a6b4.

**Comment by pllim** [comment]
In the interest of getting something in quickly to unblock downstream CI against dev astropy, I propose we just do it the way we are already doing it (dynamically) as @mhvk mentioned above. If this PR is not update by 12 PM US Eastern NYC timezone (I know it is getting late in Europe), I will step in with an alternative PR but I would still credit @neutrinoceros in the commit as co-author.

**Comment by pllim** [comment]
If you feel strongly about hardcoding the names, please open a follow-up issue to discuss because a lot of stuff will need to be changed, not just this line.

**Comment by neutrinoceros** [comment]
I didn't realise this was blocking. I'll update the PR now with your preferred solution @pllim !

**Comment by dhomeier** [comment]
I've just committed the alternative solution, as there have obviously been no hard feelings against this expressed, and @mhvk was also leaning towards this.

**Comment by pllim** [comment]
Thanks for the quick response, you two! I will approve once devdeps is green unless @mhvk beats me to it. 😸 

> I didn't realise this was blocking.

No worries. It was not communicated as such before but I am starting to see downstream failures caused by this in the past day or two.

**Review by pllim — APPROVED** [review]

**Pull request merged**
